### PR TITLE
file-loader/file generating plugin support

### PR DESCRIPTION
**Pull request by infinateone**
Hi matrus!

This time I have an update for all the file-loader/file gen plugin users out there.
Any assets output to a dir other than the webpack config's output dir

Example
 
output dir:                               dist/bundle
file-loader produced output: dist/assets/someimage.png
fileName:                                ../assets/someimage.png

with this change it uploads all output dir contents to the main bucket, and then all differently pathed assets go to their expected respective folders

in the example above that would be main-bucket/assets/*imagefiles.ext*

What do you think?

### CODE REVIEW

**Comment by matrus2** [comment]
Looks ok, but to verify I would like to write integration test. Could you provide a minimal webpack configuration with file-loader so I can easily reproduce the issue?

**Comment by infinateone** [comment]
Hey!

Just got back from vacation this week. I’ll get the tests made soon.

**Comment by infinateone** [comment]
Got a test made.
Updated some test utils to take the alternate pathing without affecting the existing tests.
I get failures when it tries to clear out the tmp dir

Error: ENOTEMPTY: directory not empty, rmdir 'C:\Users\ltm\Documents\GitHub\webpack-s3-uploader\test\.tmp/styles'

but that does it with or without my changes.

I commented that out to get my test running, but then ran into

Error: S3Plugin-RequiredS3UploadOpts: Bucket

on all the tests, but I believe you are aware of the issue so that should be fine.

I verified the build output folder structure is correct.

Let me know if you need anything else.

**Comment by infinateone** [comment]
Hi matrus!

You get a chance to take a look? No hurry, but whenever you get a chance let me know if you need anything.

**Comment by matrus2** [comment]
Hey, I just got back from a vacation, will look this week.

File: test/upload_test.js [hunk 1]
```diff
@@ -27,6 +27,14 @@ describe('S3 Webpack Upload', () => {
         .then(assertFileMatches);
     });
 
+    it('uploads entire directory with alternate pathing s3', () => { // eslint-disable-line
+      config = testHelpers.createWebpackConfigAlt({ s3Config });
```
Comment (matrus2) [inline test/upload_test.js]: Why not put this test within its own `describe` function as for now `config` param shadows previously generated config?

File: test/upload_test.js [hunk 2]
```diff
@@ -27,6 +27,14 @@ describe('S3 Webpack Upload', () => {
...
+
+      return testHelpers.runWebpackConfig({ config })
+        .then(testHelpers.testForFailFromDirectoryOrGetS3Files(testHelpers.OUTPUT_PATH_ALT))
```
Comment (matrus2) [inline test/upload_test.js]: I see two things here:
1. You pass `OUTPUT_PATH_ALT`, which is `bundle` directory, but image file is not there. It is in `assets` directory. So this test is not testing entire functionality.
2. Function `testForFailFromDirectoryOrGetS3Files` at the end of the chain invokes `expected: this.readFileFromOutputDir(fetchFile),`, which uses OUTPUT_PATH. That is why test fails - there are no files within `OUTPUT_PATH`. My suggestion is to remove entirely `OUTPUT_PATH_ALT`.

File: test/upload_test_helpers.js [hunk 3]
```diff
@@ -13,10 +13,15 @@ const S3_URL = s3Opts.AWS_S3_URL;
 const S3_ERROR_REGEX = /<Error>/;
 const OUTPUT_FILE_NAME = 's3Test';
 const OUTPUT_PATH = path.resolve(__dirname, '.tmp');
+const OUTPUT_PATH_ALT = path.join(__dirname, '.tmp/bundle');
 const FIXTURES_PATH = path.resolve(__dirname, 'fixtures');
 const ENTRY_PATH = path.resolve(__dirname, 'fixtures/index.js');
 const createBuildFailError = errors => `Webpack Build Failed ${errors}`;
 
+const ASSET_FILE_NAME_PATTERN = '[name]@[hash].[ext]';
+const ASSET_PATH= '/assets';
```
Comment (matrus2) [inline test/upload_test_helpers.js]: Please fix eslint errors.

File: test/upload_test_helpers.js [hunk 4]
```diff
@@ -92,11 +98,14 @@ module.exports = {
 
   cleanOutputDirectory() {
     deleteFolderRecursive(OUTPUT_PATH);
+	deleteFolderRecursive(OUTPUT_PATH_ALT);
   },
 
-  createOutputPath() {
-    if (!fs.existsSync(OUTPUT_PATH)) {
-      fs.mkdirSync(OUTPUT_PATH);
+  createOutputPath(outputPath) {
```
Comment (matrus2) [inline test/upload_test_helpers.js]: You can set default values in the following way `createOutputPath(output = OUTPUT_PATH)`

File: test/upload_test_helpers.js [hunk 5]
```diff
@@ -134,8 +143,31 @@ module.exports = {
...
+        filename: `${OUTPUT_FILE_NAME}-[hash]-${+new Date()}.js`,
+      },
+    }, config);
+  },
+
   runWebpackConfig({ config }) {
-    this.createOutputPath();
+    this.createOutputPath(config.output.path);
 
     return new Promise((resolve, reject) => {
```
Comment (matrus2) [inline test/upload_test_helpers.js:167]: It would be good if you could add comprehensive description in README.md

**Review by matrus2 — CHANGES_REQUESTED** [review]
Fix tests
Add desc to readme

**Comment by infinateone** [comment]
Thanks for the comments! Will get it done as soon as I can.

File: test/upload_test_helpers.js (hunk 5 above)
Comment (infinateone) [reply test/upload_test_helpers.js:167]: Updated readme

**Comment by infinateone** [comment]
Knocking out the comments as I get time. I am using your CR comments to track progress and will be updating when I get them complete locally, and I will put up the finished CR when all are addressed. Almost done! Thanks for your patience!

File: test/upload_test_helpers.js (hunk 4 above)
Comment (infinateone) [reply test/upload_test_helpers.js]: Switched to param defaulting

File: test/upload_test_helpers.js (hunk 3 above)
Comment (infinateone) [reply test/upload_test_helpers.js]: Added space before =

File: test/upload_test.js (hunk 1 above)
Comment (infinateone) [reply test/upload_test.js]: Moved to its own describe function

File: test/upload_test.js (hunk 2 above)
Comment (infinateone) [reply test/upload_test.js]: Removed OUTPUT_PATH_ALT in favor of the regular OUTPUT_PATH

File: test/upload_test_helpers.js (hunk 3 above)
Comment (infinateone) [reply test/upload_test_helpers.js]: Addressed the rest of the lint issues

**Comment by infinateone** [comment]
I think I got everything. Let me know if you need anything else.

**Comment by matrus2** [comment]
@infinateone Sorry for checking this with delay. I have only two very minor things and we can merge it.

File: README.md [hunk 6]
```diff
@@ -71,6 +71,58 @@ It is required to set:
 `include` and `exclude` rules behave similarly to Webpack's loader options.  In addition to a RegExp you can pass a function which will be called with the path as its first argument.  Returning a truthy value will match the rule.  You can also pass an Array of rules, all of which must pass for the file to be included or excluded.
 
 
+##### Pathing for resources outside of output.path
+Resources that are located outside of the webpack output.path directory are pathed as follows
```
Comment (matrus2) [inline README.md:75]: I would change this to be super super clear:
`Resources that are compiled during webpack build outside of the webpack output.path directory are pathed as follows`.

File: test/upload_test.js [hunk 7]
```diff
@@ -40,6 +40,25 @@ describe('S3 Webpack Upload', () => {
     });
   });
 
+  describe('With directory and alternate pathing', () => {
+    let s3Config;
+    let config;
+
+    beforeEach(() => {
+      config = testHelpers.createWebpackConfig({ s3Config });
```
Comment (matrus2) [inline test/upload_test.js:48]: Please remove this one. It is redundant.

**Review by matrus2 — APPROVED** [review]

**Comment by infinateone** [comment]
Thank you so much! I will get on it very soon.

**Comment by matrus2** [comment]
@infinateone Disregard my comments, I already merged your changes.

**Comment by matrus2** [comment]
v1.3.0

**Pull request merged**